File: Sources/YTools/Services/UsageRankingStore.swift
import CryptoKit
import Foundation

/// Stores only SHA-256 hashes of result identifiers, so usage learning does not
/// create a second plaintext index of private file paths.
final class UsageRankingStore {
    private struct Entry: Codable {
        var count: Int
        var lastUsed: Date
    }

    private struct Latch: Codable {
        var resultHash: String
        var confirmations: Int
        var lastUsed: Date
    }

    private struct StoreData: Codable {
        var version = 2
        var entries: [String: Entry]
        var latches: [String: Latch]
    }

    private let fileURL: URL
    private var entries: [String: Entry]
    private var latches: [String: Latch]

    init(fileManager: FileManager = .default) {
        let applicationSupport = fileManager.urls(
            for: .applicationSupportDirectory,
            in: .userDomainMask
        ).first ?? fileManager.homeDirectoryForCurrentUser.appendingPathComponent("Library/Application Support")
        // Compatibility namespace retained across the product rename.
        let directory = applicationSupport.appendingPathComponent("ZToolsNative", isDirectory: true)
        try? fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
        self.fileURL = directory.appendingPathComponent("usage-ranking.json")
        if let data = try? Data(contentsOf: fileURL),
           let decoded = try? JSONDecoder().decode(StoreData.self, from: data) {
            self.entries = decoded.entries
            self.latches = decoded.latches
        } else if let data = try? Data(contentsOf: fileURL),
                  let legacy = try? JSONDecoder().decode([String: Entry].self, from: data) {
            self.entries = legacy
            self.latches = [:]
        } else {
            self.entries = [:]
            self.latches = [:]
        }
        prune()
    }

    func record(_ identifier: String, query: String) {
        let key = hash(identifier)
        let old = entries[key]
        entries[key] = Entry(count: min((old?.count ?? 0) + 1, 10_000), lastUsed: Date())
        let normalizedQuery = normalize(query)
        if !normalizedQuery.isEmpty {
            let queryHash = hash(normalizedQuery)
            let oldLatch = latches[queryHash]
            let confirmations = oldLatch?.resultHash == key
                ? min((oldLatch?.confirmations ?? 0) + 1, 20)
                : 1
            latches[queryHash] = Latch(
                resultHash: key,
                confirmations: confirmations,
                lastUsed: Date()
            )
        }
        prune()
        save()
    }

    func boost(for identifier: String, query: String) -> Int {
        let resultHash = hash(identifier)
        let entry = entries[resultHash]
        let frequency = entry.map { Int(log2(Double($0.count + 1)) * 24) } ?? 0
        let age = entry.map { Date().timeIntervalSince($0.lastUsed) } ?? .greatestFiniteMagnitude
        let recency: Int
        switch age {
        case ..<3_600: recency = 45
        case ..<86_400: recency = 30
        case ..<(7 * 86_400): recency = 15
        default: recency = 0
        }
        let normalizedQuery = normalize(query)
        let latchBoost: Int
        if !normalizedQuery.isEmpty,
           let latch = latches[hash(normalizedQuery)],
           latch.resultHash == resultHash,
           Date().timeIntervalSince(latch.lastUsed) < 28 * 86_400 {
            latchBoost = min(180, 55 + latch.confirmations * 25)
        } else {
            latchBoost = 0
        }
        return frequency + recency + latchBoost
    }

    func clear() {
        entries.removeAll()
        latches.removeAll()
        try? FileManager.default.removeItem(at: fileURL)
    }

    private func hash(_ value: String) -> String {
        SHA256.hash(data: Data(value.utf8)).map { String(format: "%02x", $0) }.joined()
    }

    private func normalize(_ query: String) -> String {
        query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }

    private func prune() {
        let cutoff = Date().addingTimeInterval(-28 * 86_400)
        entries = entries.filter { $0.value.lastUsed >= cutoff }
        latches = latches.filter { $0.value.lastUsed >= cutoff }
    }

    private func save() {
        let payload = StoreData(entries: entries, latches: latches)
        guard let data = try? JSONEncoder().encode(payload) else { return }
        do {
            try data.write(to: fileURL, options: .atomic)
            try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: fileURL.path)
        } catch {
            // Ranking is optional and must never prevent launching a result.
        }
    }
}

File: src/YTools.Windows/Services/Storage/UsageRankingStore.cs
using System.IO;
using System.Security.Cryptography;
using System.Text.Json;
using YTools.Infrastructure;

namespace YTools.Services.Storage;

/// <summary>
/// Stores only SHA-256 hashes of result identifiers, so usage learning does
/// not create a second plaintext index of private file paths.
/// </summary>
public sealed class UsageRankingStore
{
    private readonly object _lock = new();
    private readonly string _filePath;
    private Dictionary<string, Entry> _entries;
    private Dictionary<string, Latch> _latches;

    public UsageRankingStore()
    {
        AppPaths.EnsureDirectories();
        _filePath = AppPaths.UsageRankingFile;
        (_entries, _latches) = Load();
        Prune();
    }

    public void Record(string identifier, string query)
    {
        lock (_lock)
        {
            var key = Hash(identifier);
            _entries[key] = new Entry(
                Math.Min((_entries.TryGetValue(key, out var old) ? old.Count : 0) + 1, 10_000),
                DateTimeOffset.UtcNow);
            var normalizedQuery = Normalize(query);
            if (!string.IsNullOrEmpty(normalizedQuery))
            {
                var queryHash = Hash(normalizedQuery);
                var confirmations = _latches.TryGetValue(queryHash, out var oldLatch)
                    && oldLatch.ResultHash == key
                    ? Math.Min(oldLatch.Confirmations + 1, 20)
                    : 1;
                _latches[queryHash] = new Latch(key, confirmations, DateTimeOffset.UtcNow);
            }

            Prune();
            Save();
        }
    }

    public int Boost(string identifier, string query)
    {
        lock (_lock)
        {
            var resultHash = Hash(identifier);
            var frequency = _entries.TryGetValue(resultHash, out var entry)
                ? (int)(Math.Log2(entry.Count + 1) * 24)
                : 0;
            var age = _entries.TryGetValue(resultHash, out var entry2)
                ? DateTimeOffset.UtcNow - entry2.LastUsed
                : TimeSpan.MaxValue;
            var recency = age < TimeSpan.FromHours(1) ? 45
                : age < TimeSpan.FromDays(1) ? 30
                : age < TimeSpan.FromDays(7) ? 15
                : 0;

            var normalizedQuery = Normalize(query);
            var latchBoost = 0;
            if (!string.IsNullOrEmpty(normalizedQuery)
                && _latches.TryGetValue(Hash(normalizedQuery), out var latch)
                && latch.ResultHash == resultHash
                && DateTimeOffset.UtcNow - latch.LastUsed < TimeSpan.FromDays(28))
            {
                latchBoost = Math.Min(180, 55 + latch.Confirmations * 25);
            }

            return frequency + recency + latchBoost;
        }
    }

    public void Clear()
    {
        lock (_lock)
        {
            _entries.Clear();
            _latches.Clear();
            try
            {
                File.Delete(_filePath);
            }
            catch
            {
                // Ranking is optional and must never prevent launching a result.
            }
        }
    }

    private void Prune()
    {
        var cutoff = DateTimeOffset.UtcNow.AddDays(-28);
        _entries = _entries.Where(pair => pair.Value.LastUsed >= cutoff)
            .ToDictionary(pair => pair.Key, pair => pair.Value);
        _latches = _latches.Where(pair => pair.Value.LastUsed >= cutoff)
            .ToDictionary(pair => pair.Key, pair => pair.Value);
    }

    private void Save()
    {
        try
        {
            var payload = new StoreData
            {
                Entries = _entries,
                Latches = _latches
            };
            var json = JsonSerializer.SerializeToUtf8Bytes(payload);
            File.WriteAllBytes(_filePath, json);
            AppPaths.RestrictFile(_filePath);
        }
        catch
        {
            // Ranking is optional.
        }
    }

    private (Dictionary<string, Entry>, Dictionary<string, Latch>) Load()
    {
        if (!File.Exists(_filePath))
        {
            return ([], []);
        }

        try
        {
            var json = File.ReadAllBytes(_filePath);
            var data = JsonSerializer.Deserialize<StoreData>(json);
            if (data is not null)
            {
                return (data.Entries, data.Latches);
            }
        }
        catch
        {
            // Fall through to legacy or empty.
        }

        try
        {
            var json = File.ReadAllBytes(_filePath);
            var legacy = JsonSerializer.Deserialize<Dictionary<string, Entry>>(json);
            return legacy is null ? ([], []) : (legacy, []);
        }
        catch
        {
            return ([], []);
        }
    }

    private static string Hash(string value)
    {
        return Convert.ToHexString(SHA256.HashData(System.Text.Encoding.UTF8.GetBytes(value))).ToLowerInvariant();
    }

    private static string Normalize(string query)
    {
        return query.Trim().ToLowerInvariant();
    }

    private sealed record Entry(int Count, DateTimeOffset LastUsed);

    private sealed record Latch(string ResultHash, int Confirmations, DateTimeOffset LastUsed);

    private sealed class StoreData
    {
        public int Version { get; set; } = 2;

        public Dictionary<string, Entry> Entries { get; set; } = [];

        public Dictionary<string, Latch> Latches { get; set; } = [];
    }
}
